Generate/validate a PIN for our virtual smartcard

zmb3 opened the pull request:
Instead of using a shared static PIN for all sessions, use a
cryptographically secure random number generator to generate
a random 8-digit PIN that is unique per session.

Additionally, implement the verify command by checking that the
PIN that Windows sends back matches the one we randomly generated.

Since the PIN is only-known to Teleport, this prevents users from
using the virtual smartcard for anything besides the initial
Windows login.

zmb3 committed:
Generate a random smart card PIN per-session

Instead of using a shared static PIN for all sessions, use a
cryptographically secure random number generator to generate
a random 8-digit PIN that is unique per session.

Additionally, implement the verify command by checking that the
PIN that Windows sends back matches the one we randomly generated.

Since the PIN is only-known to Teleport, this prevents users from
using the virtual smartcard for anything besides the initial
Windows login.

zmb3 committed:
Fix hang if PIN doesn't match

Now that we're verifying the smart card PIN, the RDP connection between
Windows Desktop Service will be closed if the PIN doesn't match.

Unfortunately, the goroutine that's reading user input from the browser
is still blocked waiting on user input, and doesn't know that we're no
longer connected to the desktop. This situation resolves itself as soon
as you try to move the mouse or press a key, but it would be better to
shut the TDP connection down immediately rather than leave it open with
a static screen.

timothyb89 commented on lib/srv/desktop/rdp/rdpclient/src/scard.rs:
```diff
@@ -177,7 +179,13 @@ impl Client {
+            self.uuid,
+            &self.cert_der,
+            &self.key_der,
+            self.pin.clone(),
```
I don't think it's worthwhile to eliminate this `.clone()` but as a mild stye nitpick, I usually prefer to keep owned types to structs and (only where necessary) constructors. In other words, pass the pin as a `&str` to `ctx.connect()` and call `.clone()` on that as late as possible (ideally as the struct is being built).

This helps make sure we don't create unnecessary clones elsewhere as things get passed around.

espadolini replied:
The Rust API guidelines disagree with this: https://rust-lang.github.io/api-guidelines/flexibility.html#caller-decides-where-to-copy-and-place-data-c-caller-control
> If a function requires ownership of an argument, it should take ownership of the argument rather than borrowing and cloning the argument.

zmb3 replied:
I think Tim is pointing out that connect here doesn't actually require ownership of the pin because it just passes it through to the piv::Card.

codingllama commented on lib/srv/desktop/rdp/rdpclient/client.go:
```diff
@@ -243,6 +243,12 @@ func (c *Client) start() {
+			// close the TDP connection to the browser
+			// (without this the input streaming goroutine will hang
+			// waiting for user input)
+			c.cfg.Conn.SendError("There was an error reading data from the Windows Desktop")
```
Any chance we can cover this condition with a test?

zmb3 replied:
I'll have to think about it. I only discovered this corner case when I intentionally (and temporarily) modified the Rust to fail in a spot where it was previously impossible to fail.

zmb3 replied:
Yeah, covering this with a Go test would require building and linking to a Rust library that simulates the failure. Would be a heavy lift, so I think I'll put this in the bucket of "testing cgo is really hard."

codingllama replied:
Fair enough.

codingllama commented on lib/srv/desktop/tdp/conn.go:
```diff
@@ -29,6 +31,12 @@ import (
 type Conn struct {
 	rw   io.ReadWriter
 	bufr *bufio.Reader
+	once sync.Once
```
Maybe:

```suggestion
	closeOnce sync.Once
```

or add a comment explaining what the `sync.Once` is used for.

codingllama commented on lib/srv/desktop/tdp/conn.go:
```diff
@@ -29,6 +31,12 @@ import (
 	bufr *bufio.Reader
+	once sync.Once
+
+	// not required, but can be provided in order to
+	// implement srv.TrackingConn
```
```suggestion
	// Local and Remote are not required, but can be provided in order to
	// implement srv.TrackingConn
```

codingllama commented on lib/srv/desktop/windows_server.go:
```diff
@@ -557,36 +557,40 @@ func (s *WindowsService) Serve(plainLis net.Listener) error {
-	defer proxyConn.Close()
 	log := s.cfg.Log
+
 	tdpConn := tdp.NewConn(proxyConn)
```
Suggestion:

```suggestion
	tdpConn := tdp.NewTrackingConn(proxyConn)
```

where NewTrackingConn is:

```go
type trackingConn interface {
	io.ReadWriter
	srv.TrackingConn
}

func NewTrackingConn(c trackingConn) *Conn { ... }
```

We could even make the Local and Remote fields private, or make separate conn types (as suggested in another comment).

zmb3 replied:
Not a huge fan of an exported constructor that accepts an unexported interface as an argument..

zmb3 replied:
Ohh I know what I'll do.

- I'll keep the signature the same.
- I'll unexport local/remote fields
- In the constructor I'll do a type assertion on the readwriter you provided, and if it has a `LocalAddr()` and `RemoteAddr()` method I will set the private fields.

This way, you get a tracking conn if you're wrapping a tracking conn. And if you're wrapping something like a bytes.Buffer for tests, you'll just get empty local/remote addrs.

codingllama replied:
Sure, sounds alright. Friendly reminder to note this behavior in the comments.

codingllama commented on lib/srv/desktop/tdp/conn.go:
```diff
@@ -57,3 +76,18 @@ func (c *Conn) OutputMessage(m Message) error {
+}
+
+// LocalAddr returns local address
+func (c *Conn) LocalAddr() net.Addr {
```
What happens if a Conn without Local and Remote is used in place of a TrackingConn? Should we use separate constructors and types so it can't be misused?

zmb3 replied:
I made `tdp.Conn` implement `srv.TrackingConn` with these fields so we could use the `tdp.Conn` directly with our idle connection monitor.

Prior to this change, the `tdp.Conn` wrapped a `tls.Conn`, and we passed the underlying `tls.Conn` to the connection monitor. This makes it hard to do a nice clean shutdown as the monitor would close the `tls.Conn` out from underneath you, before you had a chance to notify the end user why the connection is being closed.

If you don't specify the local/remote fields, the `tdp.Conn` still parses messages and runs a remote desktop session as usual, but places that expect a `TrackingConn` will omit those fields in any audit events that they emit.

We could make them required, but I kind of like that a `tdp.Conn` just wraps an arbitrary `ReadWriter` - nice for testing without making a network connection.

codingllama replied:
So sounds like a TrackingConn is a nice-to-have, but it won't break anything if a non-tracking one is used. Fair enough.

codingllama commented on lib/srv/desktop/windows_server.go:
```diff
@@ -557,36 +557,40 @@ func (s *WindowsService) Serve(plainLis net.Listener) error {
 	if atomic.LoadInt32(&s.ldapInitialized) != 1 {
-		// TODO(zmb3): send TDP error message
-		log.Error("This service cannot accept connections until LDAP initialization has completed.")
+		msg := "This service cannot accept connections until LDAP initialization has completed."
```
nit:

```suggestion
		const msg = "This service cannot accept connections until LDAP initialization has completed."
```

codingllama reviewed (commented):
Go review.

zmb3 committed:
Don't error when PIN verify fails

By returning an error, we kill the connection.
Instead, just return the error that indicates verification failed.
Windows won't allow access to any resources, but the connection will
remain open.
